plugins/inputs/suricata/suricata.go

```diff
@@ -0,0 +1,215 @@
+		s.InputListener.SetUnlinkOnClose(true)
+		go s.handleServerConnection(acc)
+	}
+	return nil
+}
+
+// Stop causes the plugin to cease collection JSON data from the socket provided
+// to Suricata.
+func (s *Suricata) Stop() {
+	s.Lock()
+	defer s.Unlock()
+	if s.CloseChan != nil {
+		s.InputListener.Close()
+		s.ClosedChan = make(chan bool)
+		close(s.CloseChan)
+		<-s.ClosedChan
+		s.CloseChan = nil
+	}
+}
+
+func (s *Suricata) handleServerConnection(acc telegraf.Accumulator) {
+	var err error
+	for {
+		select {
+		case <-s.CloseChan:
+			s.InputListener = nil
+			close(s.ClosedChan)
+			return
+		default:
+			var conn net.Conn
+			conn, err = s.InputListener.Accept()
+			if err != nil {
+				continue
+			}
+			reader := bufio.NewReaderSize(conn, 10485760)
+		out:
+			for {
+				select {
+				case <-s.CloseChan:
+					conn.Close()
+					s.InputListener.Close()
+					s.InputListener = nil
+					close(s.ClosedChan)
+					return
+				default:
+					line, isPrefix, rerr := reader.ReadLine()
+					if rerr == nil {
+						if isPrefix {
+							acc.AddError(errors.New("incomplete line read from input"))
+							continue
+						} else {
+							s.parse(acc, line)
+						}
+					} else if rerr == io.EOF {
+						break out
+					}
+				}
+			}
+		}
+	}
+}
+
+func splitAtSingleDot(in string) []string {
+	res := singleDotRegexp.FindAllStringIndex(in, -1)
+	if res == nil {
+		return []string{in}
+	}
+	ret := make([]string, 0)
+	startpos := 0
+	for _, v := range res {
+		ret = append(ret, in[startpos:v[0]+1])
+		startpos = v[1] - 1
+	}
+	return append(ret, in[startpos:])
+}
+
+func addPercentage(thread string, val1 string, val2 string, valt string,
+	fields map[string](map[string]interface{})) {
+	if _, ok := fields[thread][val1]; ok {
+		if _, ok := fields[thread][val2]; ok {
+			f1 := fields[thread][val1].(float64)
+			f2 := fields[thread][val2].(float64)
+			fields[thread][valt] = f2 / (f1 + f2)
+		}
+	}
+}
+
+func (s *Suricata) parse(acc telegraf.Accumulator, sjson []byte) {
+	if len(sjson) == 0 {
+		return
+	}
+
+	out, err := gojsonexplode.Explodejsonstr(string(sjson), ".")
+	if err != nil {
+		acc.AddError(err)
+		return
+	}
+
+	var result map[string]interface{}
+	err = json.Unmarshal([]byte(out), &result)
+	if err != nil {
+		acc.AddError(err)
+		return
+	}
+
+	fields := make(map[string](map[string]interface{}))
+
+	for k, v := range result {
+		key := strings.Replace(k, "stats.", "", 1)
+		if strings.HasPrefix(key, "threads.") {
+			key = strings.Replace(key, "threads.", "", 1)
+			threadkeys := splitAtSingleDot(key)
+			threadkey := threadkeys[0]
+			if _, ok := fields[threadkey]; !ok {
+				fields[threadkey] = make(map[string]interface{})
+			}
+			fields[threadkey][strings.Join(threadkeys[1:], ".")] = v
+		} else {
+			if _, ok := fields["total"]; !ok {
+				fields["total"] = make(map[string]interface{})
+			}
+			fields["total"][key] = v
+		}
+	}
+
+	addPercentage("total", "capture.kernel_packets", "capture.kernel_drops",
```

Comment:
We should probably remove these, we don't want to do any real processing in the input and instead leave it for processors or query time.

Reply:
Removed.